### ISSUE TITLE
Inconsistent behaviour with preprocessing(bandpass) of frames

### ISSUE THREAD

**Issue by rfhigler** [soft-matter/trackpy#355]
The following to code snippets give entirely different results even though functionally they are identical. The only difference is the whether I call tp.bandpass() before I do tp.locate() or whether I use tp.locate() to call tp.bandpass() for me.

`frame_bp = tp.bandpass(frames[0], [1, 1, 1], [13, 13, 13], threshold=None, truncate=4)`
`test_loc2 = tp.locate(frame_bp, diameter=(13,13,13), separation=(9,9,9), engine='numba',preprocess=False)`

`test_loc = tp.locate(frames[0], diameter=(13,13,13), separation=(9,9,9), engine='numba', noise_size=[1,1,1], smoothing_size=[13,13,13], threshold=None)`

The first method gives the expected results and find all features perfectly, while the bottom approach mainly tries to locate noise. All frames are 3D stacks with a shape of (307,307,101)(x,y,z), are uint16s and have a couple of large, bright white, and spherical features against a black background. Perfect locating material. [Example slice](http://imgur.com/r8vlWNz).

I believe the bug is located in file feature.py, line 567:
`if np.issubdtype(raw_image.dtype, np.integer):` which if corrected to `if np.issubdtype(image.dtype, np.integer):` does no longer produce these inconsistent results. 

Cheers,
Ruben

**Comment by tacaswell** [comment]
@rfhigler Can you put in a PR making that change?

**Comment by rfhigler** [comment]
@tacaswell Yes I could, but after looking a bit further I don't think my changes actually fixes anything. They only work in my very specific case where I start with uint16s, bandpass them into float64 and rescale them back to either uint8 or uint16. Going the default route (uint16) results in ~3300 local maxima while I only expect 86. The uint8 route works fine; but forcing a rescale of every input to uint8s seems a bit of a hack to me. 

I can get rid of all these spurious maxima hits by increasing the bandpass 'threshold' parameter from its default value of 1 (for integers) to about 2500. I have never had this issue with Maria Kilfoil's matlab/python code, but her code does use a different algorithm to find the local maxima (it doesn't use dilation).

Anyway, as far as I'm concerned this issue can be closed since I can solve it by increasing the threshold value manually. If anyone is interested in reproducing my findings or figuring out exactly what's what I would be more than happy to share my image-data and ipython notebooks.

Cheers,
Ruben

**Comment by caspervdw** [comment]
Hi @rfhigler, thanks for your issue report.

After `locate` performs the bandpass, it always normalizes the image and coerces it to integer type. Which bitsize it uses, depends on the input image: if the original image was of float type, it will be an uint8; if the original image was of integer type, it will take that datatype (for you: uint16). That is what line 567 is about. If you do the preprocessing yourself, trackpy does not know which original dtype it was, and it will always normalize to uint8. This could give minor differences, but not as large as you describe.

A second issue is the intensity-based feature filtering (parameter `minmass`). This parameter should be change accordingly with your input image intensity: for float images this would be much lower than for integer images. There is always a default value set: for integer values, it is 100, for float values it is 1. I think this is what you should tune.

If you turn off intensity-based filtering (`minmass=0`), you could in principle reproduce your results if you do the normalizing the same as `locate` does (use `trackpy.preprocessing.scale_to_gamut`). If you use mass-based filtering, there is no way to exactly reproduce the result with custom preprocessing, because `minmass` is defined differently.

@danielballan I am :+1: on changing the current default behaviour of minmass, as it really obscures these kind of things. Also we should really take a look to issue #257, maybe we can put the rescaling in `bandpass` or in `local_maxima`.
